### ISSUE TITLE
What's wrong with latest mk-ca-bundle.pl?

### ISSUE THREAD

**Issue by powerman** [powerman/perl-Crypt-MatrixSSL3#21]
- Old ca-certificates.crt bundled with original 3.7.3 contains **153** certificates and successfully validates certificate on https://google.com/.
- If I run `mk-matrixssl-ca-certificates.pl` it will produce ca-certificates.crt with **133** certificates (20 root certificates expired/revoked and not replaced with new ones???), which also successfully validates certificate on https://google.com/.
- If I update `mk-ca-bundle.pl` to latest version (ours version dated 2008) and run `mk-matrixssl-ca-certificates.pl` it will produce ca-certificates.crt with **148** certificates (which looks much better), but it **fails to validate certificate on https://google.com/**!

For now I've restored old `mk-ca-bundle.pl` and add into repo just-updated ca-certificates.crt with **133** certificates. But I suppose something is wrong here, @gitcnd or @vonydev - can you help with this issue?

**Comment by vonydev** [comment]
- mk-ca-bundle.pl builds ca-certificates.crt
- mk-matrixssl-ca-certificates.pl tests if matrixSSL cand load every certificate in ca-certificates.crt one by one. If it's ok we'll keep it, if not we discard it.

This is done because matrixSSL only works with x509 v3 certificates. Now if you present matrixSSL with a bundle with n certificates and one of them doesn't pass tests, instead of just ignoring that one certificate matrixSSL stops the whole process and considers the entire bundle invalid.

Bottom line: every certificate in ca-certificates.crt must work with matrixSSL and mk-matrixssl-ca-certificates.pl makes sure of this.

**Comment by powerman** [comment]
Thanks, but I know all of this. Question is what's wrong with both old mk-ca-bundle.pl (too many certificates rejected by MatrixSSL, maybe old mk-ca-bundle.pl just don't download all possible certificates?) and new mk-ca-bundle.pl (looks like it downloads more certificates because more accepted by MatrixSSL as result, but problem is no one of these accepted certificates is able to verify signature of Google's certificate).

**Comment by vonydev** [comment]
Certificate chain for google looks like this:

```
depth=3 /C=US/O=Equifax/OU=Equifax Secure Certificate Authority
verify return:1
depth=2 /C=US/O=GeoTrust Inc./CN=GeoTrust Global CA
verify return:1
depth=1 /C=US/O=Google Inc/CN=Google Internet Authority G2
verify return:1
depth=0 /C=US/ST=California/L=Mountain View/O=Google Inc/CN=*.google.com
verify return:1
```
Geotrust has 2 root certificates: one cross-signed with Equifax and one self signed.

The new mk-ca-bundle.pl doesn't add the "Equifax Secure Certificate Authority" certificate anymore.

https://knowledge.geotrust.com/support/knowledge-base/index?page=content&id=AR1426&actp=search&viewlocale=en_US&searchid=1283360269668

**Comment by powerman** [comment]
> The new mk-ca-bundle.pl doesn't add the "Equifax Secure Certificate Authority" certificate anymore.

Well, it turns out we should prove some extra param to new mk-ca-bundle.pl to solve this:
```
	-p	list of Mozilla trust purposes and levels for certificates
		to include in output. Takes the form of a comma separated
		list of purposes, a colon, and a comma separated list of
		levels. (default: SERVER_AUTH:TRUSTED_DELEGATOR)
		  Valid purposes are:
		    ALL, DIGITAL_SIGNATURE, NON_REPUDIATION,
		    KEY_ENCIPHERMENT, DATA_ENCIPHERMENT, KEY_AGREEMENT,
		    KEY_CERT_SIGN, CRL_SIGN, SERVER_AUTH, CLIENT_AUTH,
		    CODE_SIGNING, EMAIL_PROTECTION, IPSEC_END_SYSTEM,
		    IPSEC_TUNNEL, IPSEC_USER, TIME_STAMPING,
		    STEP_UP_APPROVED
		  Valid levels are:
		    ALL, TRUSTED_DELEGATOR, NOT_TRUSTED, MUST_VERIFY_TRUST,
		    TRUSTED
```
Looks like `-p ALL:TRUSTED_DELEGATOR` works good enough and let verify google's cert. But I've no idea what exactly these categories mean for security, so feel free to reopen this issue and let me know if you believe that's wrong choice and we should use some other category or shouldn't upgrade to new mk-ca-bundle.pl at all.
